pkg/web/sources.go

```diff
@@ -0,0 +1,659 @@
+		ctx.JSON(http.StatusBadRequest, gin.H{
+			"error": "missing 'name'",
+		})
+		return
+	}
+
+	// Domains to download from.
+	if jobConfig.Domains = ctx.PostFormArray("domains"); len(jobConfig.Domains) == 0 {
+		ctx.JSON(http.StatusBadRequest, gin.H{
+			"error": "missing 'domains'",
+		})
+		return
+	}
+
+	// Allow insecure download.
+	if jobConfig.Insecure, err = strconv.ParseBool(ctx.PostForm("insecure")); err != nil {
+		ctx.JSON(http.StatusBadRequest, gin.H{
+			"error": "please specify 'insecure' boolean parameter",
+		})
+		return
+	}
+
+	if jobConfig.IgnoreSignatureCheck, err = strconv.ParseBool(ctx.PostForm("ignore_signature_check")); err != nil {
+		ctx.JSON(http.StatusBadRequest, gin.H{
+			"error": "please specify 'ignore_signature_check' boolean parameter",
+		})
+		return
+	}
+
+	for _, domain := range jobConfig.Domains {
+		if domain == "" {
+			ctx.JSON(http.StatusBadRequest, gin.H{
+				"error": "don't specify empty domains",
+			})
+			return
+		}
+	}
+
+	if clientKey := ctx.PostForm("client_key"); clientKey != "" {
+		jobConfig.ClientKey = &clientKey
+	}
+
+	if clientPassphrase := ctx.PostForm("client_passphrase"); clientPassphrase != "" {
+		jobConfig.ClientPassphrase = &clientPassphrase
+	}
+
+	if ignorePattern := ctx.PostForm("ignore_pattern"); ignorePattern != "" {
+		jobConfig.IgnorePattern = &ignorePattern
+	}
+
+	if rate, ok := ctx.GetPostForm("rate"); ok {
+		rateFloat, err := strconv.ParseFloat(rate, 32)
+		if err != nil {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		} else {
+			jobConfig.Rate = &rateFloat
+		}
+	}
+
+	if worker, ok := ctx.GetPostForm("worker"); ok {
+		workerInt, err := strconv.ParseInt(worker, 10, 64)
+		if err != nil {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		} else {
+			jobConfig.Worker = int(workerInt)
+		}
+	}
+
+	header := ctx.PostFormArray("header")
+	if header == nil {
+		header = []string{}
+	}
+
+	var file *multipart.FileHeader
+	if file, err = ctx.FormFile("client_cert"); err == nil {
+		open, err := file.Open()
+		if err != nil {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+			return
+		}
+		defer open.Close()
+
+		certFile := bytes.NewBuffer(nil)
+		if _, err := io.Copy(certFile, open); err != nil {
+			ctx.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+			return
+		}
+		jobConfig.ClientCerts = certFile.Bytes()
+	}
+
+	if startRange, ok := ctx.GetPostForm("start_range"); ok {
+		startRangeTime, err := time.Parse("2006-01-02", startRange)
+		if err != nil {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		} else {
+			jobConfig.StartRange = &startRangeTime
+		}
+	}
+
+	if endRange, ok := ctx.GetPostForm("start_range"); ok {
+		endRangeTime, err := time.Parse("2006-01-02", endRange)
+		if err != nil {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		} else {
+			jobConfig.EndRange = &endRangeTime
+		}
+	}
```

Comment:
Duplicate code in `addJob`. Change the behavior of `updateJob` or refactor it into a separate function.